node/repo/fsrepo.go

```diff
@@ -423,39 +426,59 @@ func (fsr *fsLockedRepo) Close() error {
 }
 
 // Blockstore returns a blockstore for the provided data domain.
-func (fsr *fsLockedRepo) Blockstore(ctx context.Context, domain BlockstoreDomain) (blockstore.Blockstore, error) {
-	if domain != UniversalBlockstore {
-		return nil, ErrInvalidBlockstoreDomain
-	}
-
-	fsr.bsOnce.Do(func() {
-		path := fsr.join(filepath.Join(fsDatastore, "chain"))
-		readonly := fsr.readonly
-
-		if err := os.MkdirAll(path, 0755); err != nil {
-			fsr.bsErr = err
-			return
-		}
+func (fsr *fsLockedRepo) Blockstore(ctx context.Context, domain BlockstoreDomain) (blockstore.Blockstore, func() error, error) {
+	if domain != UniversalBlockstore && domain != HotBlockstore {
+		return nil, nil, ErrInvalidBlockstoreDomain
+	}
+	if domain == UniversalBlockstore {
+		fsr.bsOnce.Do(func() {
+			path := fsr.join(filepath.Join(fsDatastore, "chain"))
+			readonly := fsr.readonly
+
+			if err := os.MkdirAll(path, 0755); err != nil {
+				fsr.bsErr = err
+				return
+			}
 
-		opts, err := BadgerBlockstoreOptions(domain, path, readonly)
-		if err != nil {
-			fsr.bsErr = err
-			return
-		}
+			opts, err := BadgerBlockstoreOptions(domain, path, readonly)
+			if err != nil {
+				fsr.bsErr = err
+				return
+			}
 
-		if system.BadgerFsyncDisable {
-			opts.SyncWrites = false
-		}
+			if system.BadgerFsyncDisable {
+				opts.SyncWrites = false
+			}
 
-		bs, err := badgerbs.Open(opts)
-		if err != nil {
-			fsr.bsErr = err
-			return
-		}
-		fsr.bs = blockstore.WrapIDStore(bs)
-	})
+			bs, err := badgerbs.Open(opts)
+			if err != nil {
+				fsr.bsErr = err
+				return
+			}
+			fsr.bs = blockstore.WrapIDStore(bs)
+		})
+		return fsr.bs, nil, fsr.bsErr
+	}
+	// else domain == HotBlockstore {
+	path, err := fsr.SplitstorePath()
+	if err != nil {
+		return nil, nil, err
+	}
+	path = filepath.Join(path, "hot.badger")
+	if err := os.MkdirAll(path, 0755); err != nil {
+		return nil, nil, err
+	}
+	readonly := fsr.readonly
+	opts, err := BadgerBlockstoreOptions(HotBlockstore, path, readonly)
+	if err != nil {
+		return nil, nil, err
+	}
+	bs, err := badgerbs.Open(opts)
+	if err != nil {
+		return nil, nil, err
+	}
```

Comment:
This change is probably too much of a hack and needs some more thought.  We probably want unified guarantees across domains.  Do we even need the one time call thing here for universal bstore?  Maybe we can get rid of that.  Alternatively maybe we can reuse our already opened hotstore when we start of the splitstore instead of closing it down first.

Reply:
Went with the second option which turns out to be quite clean